### PR TITLE
Updated Watson SDK from 3.8.0 to 4.2.1 for Conversation and Discovery

### LINKED ISSUE #116: Build issue with conversation-4.2.1.jar
Hi - am getting build issues when I use version 4.2.1 of the jar, for this code. The following imports are not available. 

com.ibm.watson.developer_cloud.conversation.v1.ConversationService
com.ibm.watson.developer_cloud.discovery.v1.model.query.QueryResponse

But it works fine with version 3.8 of the jar. Pls guide me on which is the correct one to proceed with.  The grade build though points to version 3.8.

Thanks.

### PR DESCRIPTION
**Pull request by stevenpkg**
Fixes #116

### CODE REVIEW

**Comment by stevenpkg** [comment]
The new collection id using NLU is 29a02155-3523-4f45-bd9e-7af848e016aa

File: README.md [hunk 1]
```diff
@@ -94,22 +94,15 @@ To build the application:
 2. Launch the Discovery tooling  
   ![](readme_images/discovery_tooling.png)
 
-3. Create a new data collection, name it whatever you like, and select the default configuration.
+3. Create a new data collection, name it whatever you like, and select the default configuration. The default configuration now uses Natural Language Understanding ``NLU``.
```
Comment (germanattanasio) [inline README.md]: Don't use `NLU`. We don't refer to services with acronyms

File: src/main/java/com/ibm/watson/apis/conversation_with_discovery/listener/SetupThread.java [hunk 2]
```diff
@@ -116,10 +119,13 @@ public void run() {
...
       discovery.setEndPoint(Constants.DISCOVERY_URL);
       discovery.setUsernameAndPassword(userName, password);
+      
+      QueryOptions queryOptions = new QueryOptions.Builder(environmentId, collectionId).query("searchText:car tire pressure").build();
+      discovery.query(queryOptions).execute();
 
-      QueryRequest.Builder queryBuilder = new QueryRequest.Builder(environmentId, collectionId);
-      queryBuilder.query("searchText:car tire pressure");
-      discovery.query(queryBuilder.build()).execute();
+      //QueryRequest.Builder queryBuilder = new QueryRequest.Builder(environmentId, collectionId);
```
Comment (germanattanasio) [inline src/main/java/com/ibm/watson/apis/conversation_with_discovery/listener/SetupThread.java]: remove commented code

File: src/main/java/com/ibm/watson/apis/conversation_with_discovery/rest/ProxyResource.java [hunk 3]
```diff
@@ -75,7 +77,20 @@ private MessageRequest buildMessageFromPayload(InputStream body) {
           sbuilder.append("\n");
         }
       }
-      return GsonSingleton.getGson().fromJson(sbuilder.toString(), MessageRequest.class);
+      
+      MessageResponse response = GsonSingleton.getGson().fromJson(sbuilder.toString(), MessageResponse.class);
```
Comment (germanattanasio) [inline src/main/java/com/ibm/watson/apis/conversation_with_discovery/rest/ProxyResource.java]: why do we have empty lines here?

File: src/main/java/com/ibm/watson/apis/conversation_with_discovery/rest/ProxyResource.java [hunk 4]
```diff
@@ -101,13 +116,14 @@ private MessageRequest buildMessageFromPayload(InputStream body) {
...
-  private MessageResponse getWatsonResponse(MessageRequest request, String id) throws Exception {
+  private MessageResponse getWatsonResponse(MessageOptions options) throws Exception {
 
     // Configure the Watson Developer Cloud SDK to make a call to the
     // appropriate conversation service.
 
-    ConversationService service =
-        new ConversationService(API_VERSION != null ? API_VERSION : Constants.CONVERSATION_VERSION);
+    Conversation service =
+        new Conversation(API_VERSION != null ? API_VERSION : Constants.CONVERSATION_VERSION);
```
Comment (germanattanasio) [inline src/main/java/com/ibm/watson/apis/conversation_with_discovery/rest/ProxyResource.java]: Let's default to `Constants.CONVERSATION_VERSION` no need to use the ternary operator

File: src/main/java/com/ibm/watson/apis/conversation_with_discovery/utils/Constants.java [hunk 5]
```diff
@@ -26,7 +27,7 @@ private Constants() {
   public static final String CONVERSATION_URL = "https://gateway.watsonplatform.net/conversation/api";
   
   /** The Constant CONVERSATION_VERSION. */
-  public static final String CONVERSATION_VERSION = ConversationService.VERSION_DATE_2016_09_20;
+  public static final String CONVERSATION_VERSION = Conversation.VERSION_DATE_2017_05_26;
```
Comment (germanattanasio) [inline src/main/java/com/ibm/watson/apis/conversation_with_discovery/utils/Constants.java]: Let's use Strings for our default values. We are going to remove the constants from our SDkKs

File: src/test/java/com/ibm/watson/apis/conversation_with_discovery/rest/ProxyResourceTest.java [hunk 6]
```diff
@@ -113,13 +116,13 @@ public void testSendMessage() throws IOException, InterruptedException {
...
-    List<String> serviceText = serviceResponse.getText();
-    List<String> mockText = serviceResponse.getText();
+    List<String> serviceText = serviceResponse.getOutput().getText();
+    List<String> mockText = serviceResponse.getOutput().getText();
     assertNotNull(serviceText);
     assertNotNull(mockText);
     assertTrue(serviceText.containsAll(mockText) && mockText.containsAll(serviceText));
     assertEquals(serviceResponse, mockResponse);
-    assertEquals(serviceResponse.getTextConcatenated(" "), mockResponse.getTextConcatenated(" "));
+    //assertEquals(serviceResponse.getTextConcatenated(" "), mockResponse.getTextConcatenated(" "));
```
Comment (germanattanasio) [inline src/test/java/com/ibm/watson/apis/conversation_with_discovery/rest/ProxyResourceTest.java]: remove commented code

**Review by germanattanasio — CHANGES_REQUESTED** [review]
This looks pretty good @stevenpkg. I left some comments.

Can you also format the code with two spaces? I think your IDE will automatically do it if you press `CMD`+`F`.

**Review by germanattanasio — APPROVED** [review]

**Pull request merged**